hazelcast-jclouds/src/main/java/com/hazelcast/jclouds/JCloudsDiscoveryStrategy.java

```
@@ -0,0 +1,108 @@
+        this.computeServiceBuilder.build();
+    }
+
+    @Override
+    public Iterable<DiscoveredNode> discoverNodes() {
+        List<DiscoveredNode> discoveredNodes = new ArrayList<DiscoveredNode>();
+        try {
+            Iterable<? extends NodeMetadata> nodes =  computeServiceBuilder.getFilteredNodes();
+            for (NodeMetadata metadata : nodes) {
+                if (metadata.getStatus() != NodeMetadata.Status.RUNNING) {
+                    continue;
+                }
+                discoveredNodes.add(buildDiscoveredNode(metadata));
+            }
+        } catch (Exception e) {
+            throw new HazelcastException("Failed to get registered addresses", e);
+        }
+        return discoveredNodes;
+    }
+
+    @Override
+    public void destroy() {
+        computeServiceBuilder.destroy();
+    }
+
+    private DiscoveredNode buildDiscoveredNode(NodeMetadata metadata) {
+        Address privateAddressInstance = null;
+        if (!metadata.getPrivateAddresses().isEmpty()) {
+            InetAddress privateAddress = mapAddress(metadata.getPrivateAddresses().iterator().next());
+            privateAddressInstance =  new Address(privateAddress, NetworkConfig.DEFAULT_PORT);
+        }
+
+        Address publicAddressInstance = null;
+        if (!metadata.getPublicAddresses().isEmpty()) {
+            InetAddress publicAddress = mapAddress(metadata.getPublicAddresses().iterator().next());
+            publicAddressInstance =  new Address(publicAddress, NetworkConfig.DEFAULT_PORT);
+        }
+
+        return new SimpleDiscoveredNode(privateAddressInstance, publicAddressInstance);
+    }
+
+    private InetAddress mapAddress(String address) {
+        if (address == null) {
+            return null;
+        }
+        try {
+            return InetAddress.getByName(address);
+        } catch (UnknownHostException e) {
+            LOGGER.warning("Address '" + address + "' could not be resolved");
```

can we fail fast here instead of setting the discovered address to `null`?